### PR TITLE
Add new Platform Struct

### PR DESCRIPTION
**Pull request by dakotalongRH**
## What does this PR do? / Related Issues / Jira
https://issues.redhat.com/browse/OSD-25012

A new struct for Platforms is being created to better align with Go standards.  Currently, platforms are handled via String constants in a helper file.  With this new approach, Platforms will be their own struct and currently contain just an array of two Strings.

As part of the process, the old String constants in the helper file were removed as well as all code referring to them.  They will instead use the new Platform struct.

The network-verifier codebase expects Strings when comparing different platform types.  The new struct uses the String() method to return the currently expected String value for these comparisons.  For example, the new AWSHCP struct will return "hostedcluster" with AWSHCP.String().

There is also a new test suite to test each function in the new Platform package.

There should be no functional changes to the network-verifier with these changes as it's only a refactor.

## Checklist

- [X] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] I have made corresponding change to the default configuration files
- [X] I have added tests that prove my fix is effective or that my feature works
- [X] I have tested the functionality against aws, it doesn't cause any regression
- [ ] I have added execution results to the PR's readme

## Reviewer's Checklist

- [ ] (This needs to be done after technical review) I've run the branch on my local, verified that the functionality is ok

## Logs 

`[dalong@dalong-thinkpadp1gen3 integration]$ ./integration --region us-east-1 --profile osd-staging-1
2024/08/29 15:11:32 using region us-east-1
2024/08/29 15:11:32 using AZ: us-east-1a
2024/08/29 15:11:32 created VPC: vpc-0e8c8d7ca598e09c3
2024/08/29 15:11:41 enableDnsSupport is true for VPC: vpc-0e8c8d7ca598e09c3
2024/08/29 15:11:41 enableDnsHostnames is true for VPC: vpc-0e8c8d7ca598e09c3
2024/08/29 15:11:41 creating private subnet: 10.0.0.0/25
2024/08/29 15:11:41 waiting up to 1m0s for subnet to become available
2024/08/29 15:11:42 subnet available: subnet-0d86c30ed83fbaf08
2024/08/29 15:11:42 created private subnet: subnet-0d86c30ed83fbaf08
2024/08/29 15:11:42 creating firewall subnet: 10.0.0.128/25
2024/08/29 15:11:42 waiting up to 1m0s for subnet to become available
2024/08/29 15:11:42 subnet available: subnet-0a4b9d09cbbb86ea3
2024/08/29 15:11:42 created firewall subnet: subnet-0a4b9d09cbbb86ea3
2024/08/29 15:11:42 creating public subnet: 10.0.1.0/25
2024/08/29 15:11:43 waiting up to 1m0s for subnet to become available
2024/08/29 15:11:43 subnet available: subnet-024e741de4cc785ef
2024/08/29 15:11:43 created public subnet: subnet-024e741de4cc785ef
2024/08/29 15:11:45 created firewall rule group: arn:aws:network-firewall:us-east-1:277304166082:stateful-rulegroup/osd-network-verifier-rule-group
2024/08/29 15:11:45 created firewall policy: arn:aws:network-firewall:us-east-1:277304166082:firewall-policy/osd-network-verifier-firewall-policy
2024/08/29 15:11:46 created firewall: arn:aws:network-firewall:us-east-1:277304166082:firewall/osd-network-verifier-firewall
2024/08/29 15:11:46 allocated EIP address: eipalloc-029e0d8c42a2c0fc7
2024/08/29 15:11:47 created NAT gateway: nat-09d7462b30ae49ae5
2024/08/29 15:11:47 created internet gateway: igw-00350c31ceeff7431
2024/08/29 15:11:48 created private subnet route table: rtb-0fdd5ba702e6cdfc1
2024/08/29 15:11:48 associated private subnet route table: rtbassoc-0eaaf31d290be62e6
2024/08/29 15:11:48 created public subnet route table: rtb-0c83f914ab679c40a
2024/08/29 15:11:48 associated public subnet route table: rtbassoc-0789b786c96360605
2024/08/29 15:11:48 created firewall subnet route table: rtb-00a17ea8f74c6116b
2024/08/29 15:11:49 associated firewall subnet route table: rtbassoc-04bb7bf39ee18e2f9
2024/08/29 15:11:49 waiting up to 30s for internet gateway attachment to become available
2024/08/29 15:11:49 internet gateway attached: igw-00350c31ceeff7431
2024/08/29 15:11:49 route created for firewall subnet 0.0.0.0/0 --> IGW
2024/08/29 15:11:49 created internet gateway route table: rtb-00a17ea8f74c6116b
2024/08/29 15:11:50 associated internet gateway route table: rtbassoc-01060d83ea3aee090
2024/08/29 15:11:50 waiting up to 5m0s for the NAT Gateway to become available
2024/08/29 15:15:16 NAT gateway available: nat-09d7462b30ae49ae5
2024/08/29 15:15:17 route created for private subnet 0.0.0.0/0 --> NAT
2024/08/29 15:15:17 waiting up to 10m0s for the firewall to become READY
2024/08/29 15:15:17 status: PROVISIONING
2024/08/29 15:15:32 status: PROVISIONING
2024/08/29 15:16:02 status: PROVISIONING
2024/08/29 15:16:21 status: PROVISIONING
2024/08/29 15:16:54 status: PROVISIONING
2024/08/29 15:18:08 status: READY
2024/08/29 15:18:08 firewall ready: arn:aws:network-firewall:us-east-1:277304166082:firewall/osd-network-verifier-firewall
2024/08/29 15:18:09 route created for public subnet 0.0.0.0/0 --> firewall
2024/08/29 15:18:09 Starting ONV egress validation
Created security group with ID: sg-06ee69793a45a0542
Created instance with ID: i-0aa88e3227be69c54
Summary:
printing out failures:
 - egressURL error: quay.io:443

printing out exceptions preventing the verifier from running the specific test:
printing out errors faced during the execution:
2024/08/29 15:22:44 egress failure: quay.io:443
2024/08/29 15:22:44 ONV egress validation: Failure!
2024/08/29 15:22:44 cleaning up route table: rtb-0fdd5ba702e6cdfc1
2024/08/29 15:22:44 disassociated association: rtbassoc-0eaaf31d290be62e6
2024/08/29 15:22:44 deleted route table: rtb-0fdd5ba702e6cdfc1
2024/08/29 15:22:44 cleaning up route table: rtb-00a17ea8f74c6116b
2024/08/29 15:22:44 disassociated association: rtbassoc-04bb7bf39ee18e2f9
2024/08/29 15:22:45 deleted route table: rtb-00a17ea8f74c6116b
2024/08/29 15:22:45 cleaning up route table: rtb-0c83f914ab679c40a
2024/08/29 15:22:45 disassociated association: rtbassoc-0789b786c96360605
2024/08/29 15:22:45 deleted route table: rtb-0c83f914ab679c40a
2024/08/29 15:22:45 cleaning up route table: rtb-0611eda159ad362a9
2024/08/29 15:22:45 disassociated association: rtbassoc-01060d83ea3aee090
2024/08/29 15:22:46 deleted route table: rtb-0611eda159ad362a9
2024/08/29 15:22:46 deleting firewall osd-network-verifier-firewall
2024/08/29 15:22:46 waiting up to 20m0s for the firewall to be deleted
2024/08/29 15:22:46 status: DELETING
2024/08/29 15:23:01 status: DELETING
2024/08/29 15:23:29 status: DELETING
2024/08/29 15:24:14 status: DELETING
2024/08/29 15:25:36 firewall deleted: osd-network-verifier-firewall
2024/08/29 15:25:37 deleting firewall policy osd-network-verifier-firewall-policy
2024/08/29 15:25:37 waiting up to 10m0s for the firewall policy to be deleted
2024/08/29 15:25:37 status: DELETING
2024/08/29 15:25:52 status: DELETING
2024/08/29 15:26:12 firewall policy deleted: osd-network-verifier-firewall-policy
2024/08/29 15:26:12 deleting firewall rule group osd-network-verifier-rule-group
2024/08/29 15:26:12 rule group deleted: osd-network-verifier-rule-group
2024/08/29 15:26:12 deleting NAT Gateway: nat-09d7462b30ae49ae5
2024/08/29 15:26:13 waiting up to 5m0s for the NAT Gateway to be deleted
2024/08/29 15:27:23 NAT gateway deleted: nat-09d7462b30ae49ae5
2024/08/29 15:27:24 EIP released: eipalloc-029e0d8c42a2c0fc7
2024/08/29 15:27:24 detached internet gateway: igw-00350c31ceeff7431 from vpc: vpc-0e8c8d7ca598e09c3
2024/08/29 15:27:24 internet gateway deleted: igw-00350c31ceeff7431
2024/08/29 15:27:25 deleted subnet: subnet-0a4b9d09cbbb86ea3
2024/08/29 15:27:25 deleted subnet: subnet-024e741de4cc785ef
2024/08/29 15:27:25 deleted subnet: subnet-0d86c30ed83fbaf08
2024/08/29 15:27:26 deleted VPC: vpc-0e8c8d7ca598e09c3
`

### CODE REVIEW

File: pkg/data/cloud/platform.go
Comment (AlexVulaj) [inline pkg/data/cloud/platform.go:31]: I'm not sure if this was intentional or not, but having a `func (x X) String() string` method on a struct will override how it gets output during string interpolation. Not necessarily anything wrong with this, but worth noting.

In any event, I think a name like "PreferredName" instead of "String" is more clear about what the method is doing. What are your thoughts on renaming this?

File: pkg/data/cloud/platform.go
Comment (AlexVulaj) [inline pkg/data/cloud/platform.go:55]:
```suggestion
// isValid returns true if the Platform is non-empty and supported by the network verifier
```

File: pkg/data/cloud/platform.go
Comment (AlexVulaj) [inline pkg/data/cloud/platform.go:38]: Go convention would have us name this `ByName`, so it's called like `platform.ByName` - I could go either way here personally.

File: pkg/data/cloud/platform.go
Comment (AlexVulaj) [reply pkg/data/cloud/platform.go:31]: Also, is there any reason to _not_ export "names" and let consumers access it directly?

File: pkg/data/cloud/platform.go
Comment (AlexVulaj) [reply pkg/data/cloud/platform.go:55]: Unless this method is supposed to be exported of course 🙂

Comment (abyrne55) [reply pkg/data/cloud/platform.go:55]: This method should indeed be exported

File: pkg/data/cloud/platform.go
Comment (abyrne55) [reply pkg/data/cloud/platform.go:38]: I'm cool with either option as well. I'll leave it up to @dakotalongRH

Comment (dakotalongRH) [reply pkg/data/cloud/platform.go]: Agreed, changing to platform.ByName

File: pkg/data/cloud/platform.go
Comment (dakotalongRH) [reply pkg/data/cloud/platform.go:31]: Leaving the overriding String method per team discussions

File: cmd/egress/cmd.go
Comment (abyrne55) [inline cmd/egress/cmd.go:65]: Any reason why we shouldn't use/accept the Platform type directly instead of strings?

File: cmd/egress/cmd.go
Comment (abyrne55) [inline cmd/egress/cmd.go:258]: nit: change the `%v`s to `%s` and then you can drop the `.String()`s

File: cmd/egress/cmd.go
Comment (abyrne55) [inline cmd/egress/cmd.go:16]: Just a thought: any reason why we have to stick to `platform`? Would just keeping the name `cloud` be too weird?

File: cmd/egress/cmd.go
Comment (abyrne55) [inline cmd/egress/cmd.go:102]: Similarly to above, do we really need to "normalize" Platforms to string anymore? Can't we use Platforms directly?

File: pkg/verifier/aws/entry_point.go
Comment (abyrne55) [inline pkg/verifier/aws/entry_point.go:1]: This PR should modify the ValidateEgressInput struct type so that we can avoid string usage through-and-through

File: pkg/probes/legacy/machine_images.go
Comment (abyrne55) [inline pkg/probes/legacy/machine_images.go:14]: Because the `Platform` type is comparable, you can use it as a map key. IOW, the line above should look like
```go
var cloudMachineImageMap = map[platform.Platform]map[cpu.Architecture]map[string]string{
```
And then you can use `platform.AWSClassic` directly w/o the `.String()`

**Review by abyrne55 — CHANGES_REQUESTED** [review]
This is great work @dakotalongRH. Most of my feedback can be summed up as: let's use our fancy new `Platform` type to the fullest instead of strings. IOW, there should be very few (if any) explicit calls to `.String()`. Doing so will give us the type safety that was the main motivation for this refactor (see Jira card description). This includes changing the external API (i.e., the `ValidateEgressInput` struct type) to accept a `Platform` instead of a string (yes, this will result in a breaking change and will merit a new y-stream release)

File: cmd/egress/cmd.go
Comment (abyrne55) [inline cmd/egress/cmd.go:16]:
```suggestion
	"github.com/openshift/osd-network-verifier/pkg/data/cloud"
```
nit

File: examples/aws/verify_egress.go
Comment (abyrne55) [inline examples/aws/verify_egress.go:10]:
```suggestion
	"github.com/openshift/osd-network-verifier/pkg/data/cloud"
```

File: pkg/probes/curl/curl_json.go
Comment (abyrne55) [inline pkg/probes/curl/curl_json.go:13]:
```suggestion
	"github.com/openshift/osd-network-verifier/pkg/data/cloud"
```

File: pkg/probes/curl/curl_json_test.go
Comment (abyrne55) [inline pkg/probes/curl/curl_json_test.go:9]:
```suggestion
	"github.com/openshift/osd-network-verifier/pkg/data/cloud"
```

File: pkg/probes/curl/machine_images.go
Comment (abyrne55) [inline pkg/probes/curl/machine_images.go:11]:
```suggestion
// e.g., cloudMachineImageMap[cloud.AWSClassic][cpu.ArchX86]["us-east-1"]
```

File: pkg/probes/legacy/legacy.go
Comment (abyrne55) [inline pkg/probes/legacy/legacy.go:10]:
```suggestion
	"github.com/openshift/osd-network-verifier/pkg/data/cloud"
```

File: pkg/probes/legacy/machine_images.go
Comment (abyrne55) [inline pkg/probes/legacy/machine_images.go:4]:
```suggestion
	"github.com/openshift/osd-network-verifier/pkg/data/cloud"
```

File: pkg/probes/legacy/machine_images.go
Comment (abyrne55) [inline pkg/probes/legacy/machine_images.go:11]:
```suggestion
// e.g., cloudMachineImageMap[cloud.AWSClassic][cpu.ArchX86]["us-east-1"]
```

File: pkg/verifier/package_verifier.go
Comment (abyrne55) [inline pkg/verifier/package_verifier.go:35]: Let's add a docstring here similar to `Timeout`. Please mention the default value (AWSClassic)

**Review by abyrne55 — COMMENTED** [review]
Looks great so far! Suggested a few nit fixes. I'll do a full functional test once this is out of draft status

**Comment by dakotalongRH** [comment]
Here are current passing logs from the integration test:
`[dalong@dalong-thinkpadp1gen3 integration]$ ./integration -profile osd-staging-2 -region us-east-1
2024/09/19 20:07:40 using region us-east-1
2024/09/19 20:07:40 using AZ: us-east-1a
2024/09/19 20:07:41 waiting up to 10s for vpc to become available
2024/09/19 20:07:41 created VPC: vpc-0c7e54dd9c51549aa
2024/09/19 20:07:41 enableDnsSupport is true for VPC: vpc-0c7e54dd9c51549aa
2024/09/19 20:07:41 enableDnsHostnames is true for VPC: vpc-0c7e54dd9c51549aa
2024/09/19 20:07:41 creating private subnet: 10.0.0.0/25
2024/09/19 20:07:42 waiting up to 1m0s for subnet to become available
2024/09/19 20:07:42 subnet available: subnet-0fd25414a2b7724c4
2024/09/19 20:07:42 created private subnet: subnet-0fd25414a2b7724c4
2024/09/19 20:07:42 creating firewall subnet: 10.0.0.128/25
2024/09/19 20:07:42 waiting up to 1m0s for subnet to become available
2024/09/19 20:07:42 subnet available: subnet-0211a7433e0ff6da5
2024/09/19 20:07:42 created firewall subnet: subnet-0211a7433e0ff6da5
2024/09/19 20:07:42 creating public subnet: 10.0.1.0/25
2024/09/19 20:07:42 waiting up to 1m0s for subnet to become available
2024/09/19 20:07:43 subnet available: subnet-01c5747583eb1cbbc
2024/09/19 20:07:43 created public subnet: subnet-01c5747583eb1cbbc
2024/09/19 20:07:43 created firewall rule group: arn:aws:network-firewall:us-east-1:811685182089:stateful-rulegroup/osd-network-verifier-rule-group
2024/09/19 20:07:44 created firewall policy: arn:aws:network-firewall:us-east-1:811685182089:firewall-policy/osd-network-verifier-firewall-policy
2024/09/19 20:07:45 created firewall: arn:aws:network-firewall:us-east-1:811685182089:firewall/osd-network-verifier-firewall
2024/09/19 20:07:45 allocated EIP address: eipalloc-01f278c19e57c652f
2024/09/19 20:07:45 created NAT gateway: nat-0f3e4ef1478399de4
2024/09/19 20:07:45 created internet gateway: igw-0d7ea1bb744a69322
2024/09/19 20:07:46 created private subnet route table: rtb-05c97ff5953f64a75
2024/09/19 20:07:46 associated private subnet route table: rtbassoc-0184207adad74d279
2024/09/19 20:07:46 created public subnet route table: rtb-0b3d852e7bbc37e45
2024/09/19 20:07:46 associated public subnet route table: rtbassoc-0e090c4585dcf0375
2024/09/19 20:07:47 created firewall subnet route table: rtb-0ee7627a3767235a1
2024/09/19 20:07:47 associated firewall subnet route table: rtbassoc-07fc1dcc8b6cbc2c0
2024/09/19 20:07:47 waiting up to 30s for internet gateway attachment to become available
2024/09/19 20:07:47 internet gateway attached: igw-0d7ea1bb744a69322
2024/09/19 20:07:47 route created for firewall subnet 0.0.0.0/0 --> IGW
2024/09/19 20:07:47 created internet gateway route table: rtb-0ee7627a3767235a1
2024/09/19 20:07:48 associated internet gateway route table: rtbassoc-0b67fe7e5a6af1c05
2024/09/19 20:07:48 waiting up to 5m0s for the NAT Gateway to become available
2024/09/19 20:09:27 NAT gateway available: nat-0f3e4ef1478399de4
2024/09/19 20:09:28 route created for private subnet 0.0.0.0/0 --> NAT
2024/09/19 20:09:28 waiting up to 10m0s for the firewall to become READY
2024/09/19 20:09:28 status: PROVISIONING
2024/09/19 20:09:43 status: PROVISIONING
2024/09/19 20:10:06 status: PROVISIONING
2024/09/19 20:10:28 status: PROVISIONING
2024/09/19 20:12:09 status: PROVISIONING
2024/09/19 20:13:13 status: READY
2024/09/19 20:13:13 firewall ready: arn:aws:network-firewall:us-east-1:811685182089:firewall/osd-network-verifier-firewall
2024/09/19 20:13:14 route created for public subnet 0.0.0.0/0 --> firewall
2024/09/19 20:13:14 Starting ONV egress validation
Using egress URL list from https://api.github.com/repos/openshift/osd-network-verifier/contents/pkg/data/egress_lists/aws-classic.yaml?ref=main at SHA 5fd26419a0a6529a2a692fee0abe591d58542ec3
Created security group with ID: sg-08a161e6efd690532
Created instance with ID: i-0dc916b8e3cb7f335
Deleting instance with ID: i-0dc916b8e3cb7f335
Deleting security group with ID: sg-08a161e6efd690532
Summary:
printing out failures:
 - egressURL error: https://quay.io:443 (Operation timed out after 4001 milliseconds with 0 out of 0 bytes received)
 - egressURL error: https://cdn01.quay.io:443 (Operation timed out after 4000 milliseconds with 0 out of 0 bytes received)
 - egressURL error: https://cdn02.quay.io:443 (Operation timed out after 4000 milliseconds with 0 out of 0 bytes received)
 - egressURL error: https://cdn03.quay.io:443 (Operation timed out after 4000 milliseconds with 0 out of 0 bytes received)

printing out exceptions preventing the verifier from running the specific test:
printing out errors faced during the execution:
2024/09/19 20:16:59 egress failure: https://quay.io:443 (Operation timed out after 4001 milliseconds with 0 out of 0 bytes received)
2024/09/19 20:16:59 egress failure: https://cdn01.quay.io:443 (Operation timed out after 4000 milliseconds with 0 out of 0 bytes received)
2024/09/19 20:16:59 egress failure: https://cdn02.quay.io:443 (Operation timed out after 4000 milliseconds with 0 out of 0 bytes received)
2024/09/19 20:16:59 egress failure: https://cdn03.quay.io:443 (Operation timed out after 4000 milliseconds with 0 out of 0 bytes received)
2024/09/19 20:16:59 ONV egress validation: Failure!
2024/09/19 20:16:59 cleaning up route table: rtb-0ee7627a3767235a1
2024/09/19 20:16:59 disassociated association: rtbassoc-07fc1dcc8b6cbc2c0
2024/09/19 20:17:00 deleted route table: rtb-0ee7627a3767235a1
2024/09/19 20:17:00 cleaning up route table: rtb-0b3d852e7bbc37e45
2024/09/19 20:17:00 disassociated association: rtbassoc-0e090c4585dcf0375
2024/09/19 20:17:00 deleted route table: rtb-0b3d852e7bbc37e45
2024/09/19 20:17:00 cleaning up route table: rtb-05c97ff5953f64a75
2024/09/19 20:17:00 disassociated association: rtbassoc-0184207adad74d279
2024/09/19 20:17:00 deleted route table: rtb-05c97ff5953f64a75
2024/09/19 20:17:00 cleaning up route table: rtb-052b9ff4054754a0a
2024/09/19 20:17:01 disassociated association: rtbassoc-0b67fe7e5a6af1c05
2024/09/19 20:17:01 deleted route table: rtb-052b9ff4054754a0a
2024/09/19 20:17:01 deleting firewall osd-network-verifier-firewall
2024/09/19 20:17:02 waiting up to 20m0s for the firewall to be deleted
2024/09/19 20:17:02 status: DELETING
2024/09/19 20:17:17 status: DELETING
2024/09/19 20:17:43 status: DELETING
2024/09/19 20:18:07 status: DELETING
2024/09/19 20:18:28 status: DELETING
2024/09/19 20:20:02 firewall deleted: osd-network-verifier-firewall
2024/09/19 20:20:03 deleting firewall policy osd-network-verifier-firewall-policy
2024/09/19 20:20:03 waiting up to 10m0s for the firewall policy to be deleted
2024/09/19 20:20:03 status: DELETING
2024/09/19 20:20:18 status: DELETING
2024/09/19 20:20:34 firewall policy deleted: osd-network-verifier-firewall-policy
2024/09/19 20:20:34 deleting firewall rule group osd-network-verifier-rule-group
2024/09/19 20:20:34 rule group deleted: osd-network-verifier-rule-group
2024/09/19 20:20:34 deleting NAT Gateway: nat-0f3e4ef1478399de4
2024/09/19 20:20:35 waiting up to 5m0s for the NAT Gateway to be deleted
2024/09/19 20:21:50 NAT gateway deleted: nat-0f3e4ef1478399de4
2024/09/19 20:21:51 EIP released: eipalloc-01f278c19e57c652f
2024/09/19 20:21:51 detached internet gateway: igw-0d7ea1bb744a69322 from vpc: vpc-0c7e54dd9c51549aa
2024/09/19 20:21:51 internet gateway deleted: igw-0d7ea1bb744a69322
2024/09/19 20:21:52 deleted subnet: subnet-01c5747583eb1cbbc
2024/09/19 20:21:52 deleted subnet: subnet-0fd25414a2b7724c4
2024/09/19 20:21:52 deleted subnet: subnet-0211a7433e0ff6da5
2024/09/19 20:21:52 No SecurityGroups found - skipping cleanup
2024/09/19 20:21:53 deleted VPC: vpc-0c7e54dd9c51549aa`

**Comment by dakotalongRH** [comment]
Tested and verified working using the integration tests in osd-network-verifier.  Before merging, please build the binary and conduct one test to verify it's working for you as well since this is such a large code change.

**Review by abyrne55 — APPROVED** [review]
Tested locally with standalone CLI tool and two platform types. Confirmed proper operation. Excellent work, @dakotalongRH!

Note to anyone cutting the next release: this PR merits a Y-stream bump and will require downstream changes.

/approve
/lgtm

**Pull request merged**
